### PR TITLE
AB#25: Auth endpoints

### PR DESCRIPTION
**Pull request by popemkt**

### CODE REVIEW

File: EzMap.Domain/Models/PoiConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
+
+namespace EzMap.Domain;
+
+internal class PoiConfiguration :  IEntityTypeConfiguration<Poi>
+{
+    public void Configure(EntityTypeBuilder<Poi> builder)
+    {
+        builder.ToTable("Poi").HasKey(p => p.PoiId);
```
Comment (popemkt) [inline EzMap.Domain/Models/PoiConfiguration.cs]: No hard coded string

File: EzMap.Domain/Models/Poi.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿namespace EzMap.Domain;
+
+public class Poi
+{
+    public int PoiId { get; set; }
```
Comment (popemkt) [inline EzMap.Domain/Models/Poi.cs]: Only Id for the name is good, also consider Guid as type

File: EzMap.Domain/Models/User.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+﻿namespace EzMap.Domain;
+
+public class User
+{
+    public int UserId { get; set; }
+    public string? DisplayName { get; set; }
+    public string? UserName { get; set; }
+    public string? Email { get; set; }
+    public string? Password { get; set; }
```
Comment (popemkt) [inline EzMap.Domain/Models/User.cs]: Password, username, display name nullable?

Comment (bytetum) [reply EzMap.Domain/Models/User.cs]: you are right, gonna update this soon

File: EzMap.Domain/Models/PoiConfiguration.cs (hunk 1 above)
Comment (bytetum) [reply EzMap.Domain/Models/PoiConfiguration.cs]: hmm, do I need to create an enum?

File: EzMap.Domain/Models/User.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    public DateTime? CreatedDate { get; set; }
```
Comment (popemkt) [inline EzMap.Domain/Models/User.cs]: CreatedDate nullable?

File: EzMap.Domain/Models/UserConfiguration.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
+
+namespace EzMap.Domain;
+
+public class UserConfiguration : IEntityTypeConfiguration<User>
+{
+    public void Configure(EntityTypeBuilder<User> builder)
+    {
+        builder.HasNoKey();
+        builder.ToTable("User");
```
Comment (popemkt) [inline EzMap.Domain/Models/UserConfiguration.cs]: No hardcoded string, use `nameof` operator

File: EzMap.Domain/Models/UserConfiguration.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        builder.Property(u => u.UserId).HasColumnName("UserId");
+        builder.Property(u => u.DisplayName).HasMaxLength(60).IsUnicode(false);
+        builder.Property(u => u.UserName).HasMaxLength(30).IsUnicode(false);
+        builder.Property(u => u.Email).HasMaxLength(50).IsUnicode(false);
+        builder.Property(u => u.Password).HasMaxLength(20).IsUnicode(false);
```
Comment (popemkt) [inline EzMap.Domain/Models/UserConfiguration.cs]: 20 char password?

File: EzMap.Domain/Models/UserConfiguration.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        builder.Property(u => u.CreatedDate).IsUnicode(false);
```
Comment (popemkt) [inline EzMap.Domain/Models/UserConfiguration.cs]: Not everything needs Unicode settings, especially non-string type like DateTime

File: EzMap.Domain/Models/EzMapContext.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    
+    public EzMapContext(DbContextOptions<EzMapContext> options)
+        : base(options)
+    {
+    }
+    
+    
+    protected override void OnModelCreating(ModelBuilder modelBuilder)
+    {
+        modelBuilder.HasDefaultSchema("mc")
```
Comment (popemkt) [inline EzMap.Domain/Models/EzMapContext.cs]: why change default schema, and no hardcoded string

File: EzMap.Domain/Models/PoiConfiguration.cs [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+        builder.Property(p => p.Address).ValueGeneratedOnAdd();
```
Comment (popemkt) [inline EzMap.Domain/Models/PoiConfiguration.cs]: Address and Name generated?

**Comment by popemkt** [comment]
@bytetum please work according to tasks
imageI already created the first task (u should create the rest), which should only have User and Auth code
imageThat way we have clearer scopes and have smaller manageable PRs

File: EzMap.Domain/Models/User.cs (hunk 3 above)
Comment (popemkt) [reply EzMap.Domain/Models/User.cs]: No need to reply, if gonna fix it then just fix it, if need to be discussed then reply

File: EzMap.Domain/Models/UserConfiguration.cs [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace EzMap.Domain;
+
+public class UserConfiguration : IEntityTypeConfiguration<User>
+{
+    public void Configure(EntityTypeBuilder<User> builder)
+    {
+        builder.HasNoKey();
+        builder.ToTable("User");
+        builder.Property(u => u.UserId).HasColumnName("UserId");
```
Comment (popemkt) [inline EzMap.Domain/Models/UserConfiguration.cs]: Why specify column name here?
And why no Key?

File: EzMap.Domain/Repositories/PoiRepository.cs [hunk 11]
```diff
@@ -0,0 +1,104 @@
+﻿using System.Runtime.InteropServices.ComTypes;
+using Microsoft.EntityFrameworkCore;
+
+namespace EzMap.Domain.Repositories;
+
+public interface IPointOfInterests
```
Comment (popemkt) [inline EzMap.Domain/Repositories/PoiRepository.cs]: Please don't create the Repositories before you know what APIs are gonna be created, or at least don't implement the interfaces.
Some of these won't even be used.
In the end we shouldn't leave any redundant code.

File: EzMap.Domain/Repositories/PoiRepository.cs [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        try
+        {
+            Poi? poi = _dbContext.Pois.Find(id);
+            if (poi != null)
+            {
+                return poi;
+            }
+            else
+            {
+                throw new ArgumentNullException();
```
Comment (popemkt) [inline EzMap.Domain/Repositories/PoiRepository.cs]: Why `ArgumentNullException`, `poi` is not argument, `id` is

File: EzMap.Domain/Repositories/PoiRepository.cs [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+            var poi = _dbContext.Pois.Find(id);
+            if (poi != null)
+            {
+                _dbContext.Pois.Remove(poi);
+                _dbContext.SaveChanges();
+                return poi;
+            }
+            else
+            {
+                throw new ArgumentNullException();
```
Comment (popemkt) [inline EzMap.Domain/Repositories/PoiRepository.cs]: Same as above

File: ezmap-poc.sln [hunk 14]
```diff
@@ -2,17 +2,27 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "src", "src", "{41F11E7F-D94F-49D7-A341-F31DCCAFA179}"
 EndProject
-Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "backend", "backend", "{D02B54F0-A63E-4C1F-AA84-8D3E8FDAFB81}"
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "EzMap.Domain", "EzMap.Domain\EzMap.Domain.csproj", "{633AC4D5-D664-495A-8083-41E444C0D92C}"
```
Comment (popemkt) [inline ezmap-poc.sln:5]: Physical path should reflect solution structure and vice versa
i.e. physical path should be `src\EzMap.Domain\EzMap.Domain.csproj`
Same for EzMap.Api

File: EzMap.Domain/Models/PoiConfiguration.cs (hunk 1 above)
Comment (popemkt) [reply EzMap.Domain/Models/PoiConfiguration.cs]: See this comment https://github.com/Ez-Map/ezmap-poc/pull/1/files#r1337532625

File: EzMap.Domain/Repositories/PoiRepository.cs (hunk 11 above)
Comment (popemkt) [reply EzMap.Domain/Repositories/PoiRepository.cs]: Should only create minimal interface/classes while implementing.
e.g. when creating POST api, you only need `AddPoi` method, then until that point you only need `AddPoi` until you implement something else, like GET api.

An  exception is when you do something like `GenericRepository`, in that case most basic Repo methods are already implemented for every Entity.

Comment (bytetum) [reply EzMap.Domain/Repositories/PoiRepository.cs]: Are you saying that we old create a generic interface or even a base class later on, till we have all repo(s) implemented (to know what to be generic)?

Comment (popemkt) [reply EzMap.Domain/Repositories/PoiRepository.cs]: U're not doing `GenericRepo` so just forget about my latter point. Only implement what's needed is good enough.

**Comment by popemkt** [comment]
@bytetum looks good, can copy github action files from main repo for quick building

File: EzMap.Api/WeatherForecast.cs
Comment (popemkt) [inline EzMap.Api/WeatherForecast.cs:1]: Remove file

File: EzMap.Api/Controllers/PoiController.cs [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+using Microsoft.AspNetCore.Mvc;
+
+namespace EzMap.Api.Controllers;
+
+[Route("[controller]")]
+[ApiController]
+public class PoiController : ControllerBase
+{
+
+    [HttpPost("CreatePoi")]
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiController.cs]: Resource endpoints should follow REST

Comment (popemkt) [reply EzMap.Api/Controllers/PoiController.cs]: Only create?
What about update, read and delete?

File: EzMap.Domain/Migrations/20231008092541_UpdateSchema_EzMap.cs [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+        {
+            migrationBuilder.DropForeignKey(
+                name: "FK_Poi_User_UserId",
+                table: "Poi");
+
+            migrationBuilder.AlterColumn<Guid>(
+                name: "UserId",
+                table: "Poi",
+                type: "uniqueidentifier",
+                nullable: true,
```
Comment (popemkt) [inline EzMap.Domain/Migrations/20231008092541_UpdateSchema_EzMap.cs:22]: Why is this nullable?
In that case need to rethink the model and identify the source of `PoI`. When will a `PoI` is in our system:

1. Do we import all `PoI`s from external provider (e.g. gg-map) into our database? This is kind of unrealistic though. But this allows `UserId` to be `nullable`.
2. Do we save a `PoI` only when a user save it? In that case PoI must have `UserId` (not `nullable`) as the user who saved/owns it.
3. Do we allow somehow a `PoI` can be imported to our db without a user interaction? If yes then `UserId` is `nullable`, but please elaborate which scenario.

File: EzMap.Domain/Models/EntityBase.cs [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+    public Guid CreatedBy { get; set; }
+    public DateTime CreatedDate { get; set; }
+
+    public Guid? LastModifiedBy { get; set; }
+
+    public DateTime? LastModifiedDate { get; set; }
+    
+    public DateTime? DeletedDate { get; set; }
+
+    public Guid? DeletedBy { get; set; }
```
Comment (popemkt) [inline EzMap.Domain/Models/EntityBase.cs]: Currently all these properties are not populated when saving entities to database. Can reference littleviet repo if unsure how to do it.

Comment (popemkt) [inline EzMap.Domain/Models/EntityBase.cs]: DeletedBy shouldn't be needed as we can infer that from LastModifiedBy

File: EzMap.Domain/Repositories/PoiRepository.cs [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+public class PoiRepository : IPoiRepository
+{
+    private readonly EzMapContext _dbContext ;
+
+    public PoiRepository(EzMapContext dbContext)
+    {
+        _dbContext = dbContext;
+    }
+
+    public async Task ReadBooksAsync(CancellationToken token = default)
```
Comment (popemkt) [inline EzMap.Domain/Repositories/PoiRepository.cs]: the hell is readBooks??

File: EzMap.Domain/Repositories/PoiRepository.cs [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+
+public interface IPoiRepository
+{
+    Task<bool> AddPoi(PoiCreateDto dto);
+
+    Task ReadBooksAsync(CancellationToken token = default);
+}
+
+
+public class PoiRepository : IPoiRepository
```
Comment (popemkt) [inline EzMap.Domain/Repositories/PoiRepository.cs:20]: All repos should filter out "soft-deleted" entities as well.
We can do this in repo logic or use Ef's [Global query filter](https://learn.microsoft.com/en-us/ef/core/querying/filters)

File: EzMap.Api/Controllers/PoiController.cs [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+            var userId = succssful ? (Guid?)tempId : null;//TODO: fix this
+
+            uow.PoiRepository.AddPoi(dto.WithUserId(userId ?? Guid.NewGuid()));
+            
+            if (await uow.SaveAsync() > 0) return Ok("Your point of interest is created successfully!");
+
+            return new StatusCodeResult(StatusCodes.Status500InternalServerError);
+    }
+
+
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiController.cs]: Authorize

File: EzMap.Api/Controllers/PoiController.cs [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+        }
+        
+        await uow.PoiRepository.UpdatePoiAsync(dto);
+
+        if (await uow.SaveAsync() > 0) return Ok("Your point of interest is updated successfully!");
+        
+        return new StatusCodeResult(StatusCodes.Status500InternalServerError);
+    }
+
+
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiController.cs]: Authorize

File: EzMap.Api/Controllers/PoiController.cs [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+        return new StatusCodeResult(StatusCodes.Status500InternalServerError);
+    }
+
+    
+    [HttpDelete("{id:guid}")]
+    public async Task<IActionResult> Delete(Guid id, [FromServices] IUnitOfWork uow)
+    {
+        if (string.IsNullOrEmpty(id.ToString()))
+        {
+            return BadRequest("Kindly fill Name, Address fields!");
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiController.cs]: Wrong message

File: EzMap.Domain/Dtos/PoiUpdateDto.cs [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        Id = id;
+        Name = name;
+        Address = address;
+    }
+    public Guid Id { get; set; }
+    public string Name { get; set; }
+    public string Address { get; set; }
+    
+    // public Guid UserId { get; private set; }
```
Comment (popemkt) [inline EzMap.Domain/Dtos/PoiUpdateDto.cs]: add this back

File: test/EzMap.IntegrationTest/AnthenticationControllerTest.cs [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+                { Username = "redbull", Password = "strawberry" });
+
+        var content = await response.Content.ReadAsStringAsync();
+        // assert 
+        Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+        Assert.NotEmpty(content);
+    }
+
+    [Fact]
+    public async Task SignUp_ProvideCorrectData_UserRecordCreated()
```
Comment (popemkt) [inline test/EzMap.IntegrationTest/AnthenticationControllerTest.cs:40]: Correct wording

**Review by popemkt — COMMENTED** [review]
OK

**Pull request merged**